ibc/store/proofs_ics23.go

```diff
@@ -0,0 +1,159 @@
+	// verify the proof
+	return ics23.VerifyNonMembership(smtSpec, root, proof, key)
+}
+
+// createMembershipProof generates a CommitmentProof object verifying the membership of a key-value pair
+// in the SMT provided
+func createMembershipProof(tree *smt.SMT, key, value []byte) (*ics23.CommitmentProof, error) {
+	proof, err := tree.Prove(key)
+	if err != nil {
+		return nil, coreTypes.ErrCreatingProof(err)
+	}
+	return convertSMPToExistenceProof(&proof, key, value), nil
+}
+
+// createNonMembershipProof generates a CommitmentProof object verifying the membership of an unrealted key at the given key in the SMT provided
+func createNonMembershipProof(tree *smt.SMT, key []byte) (*ics23.CommitmentProof, error) {
+	proof, err := tree.Prove(key)
+	if err != nil {
+		return nil, coreTypes.ErrCreatingProof(err)
+	}
+
+	return convertSMPToExclusionProof(&proof, key), nil
+}
+
+// convertSMPToExistenceProof converts a SparseMerkleProof to an ics23
+// ExistenceProof to verify membership of an element
+func convertSMPToExistenceProof(proof *smt.SparseMerkleProof, key, value []byte) *ics23.CommitmentProof {
+	path := sha256.Sum256(key)
+	steps := convertSideNodesToSteps(proof.SideNodes, path[:])
+	return &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Exist{
+			Exist: &ics23.ExistenceProof{
+				Key:   key,
+				Value: value,
+				Leaf:  smtSpec.LeafSpec,
+				Path:  steps,
+			},
+		},
+	}
+}
+
+// convertSMPToExistenceProof converts a SparseMerkleProof to an ics23
+// ExclusionProof to verify non-membership of an element
+func convertSMPToExclusionProof(proof *smt.SparseMerkleProof, key []byte) *ics23.CommitmentProof {
+	path := sha256.Sum256(key)
+	steps := convertSideNodesToSteps(proof.SideNodes, path[:])
+	leaf := &ics23.LeafOp{
+		Hash: ics23.HashOp_SHA256,
+		// Do not re-hash already hashed fields from NonMembershipLeafData
+		PrehashKey:   ics23.HashOp_NO_HASH,
+		PrehashValue: ics23.HashOp_NO_HASH,
+		Length:       ics23.LengthOp_NO_PREFIX,
+		Prefix:       []byte{0},
+	}
+	actualPath := path[:]
+	actualValue := defaultValue
+	if proof.NonMembershipLeafData != nil {
+		actualPath = proof.NonMembershipLeafData[1:33]
+		actualValue = proof.NonMembershipLeafData[33:]
+	}
+	return &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Exclusion{
+			Exclusion: &ics23.ExclusionProof{
+				Key:             key,
+				ActualPath:      actualPath,
+				ActualValueHash: actualValue,
+				Leaf:            leaf,
+				Path:            steps,
+			},
+		},
+	}
+}
+
+// convertSideNodesToSteps converts the SideNodes field in the SparseMerkleProof
+// into a list of InnerOps for the ics23 CommitmentProof
+func convertSideNodesToSteps(sideNodes [][]byte, path []byte) []*ics23.InnerOp {
+	steps := make([]*ics23.InnerOp, 0, len(sideNodes))
+	for i := 0; i < len(sideNodes); i++ {
+		var prefix, suffix []byte
+		prefix = append(prefix, innerPrefix...)
+		if getPathBit(path, len(sideNodes)-1-i) == left {
```

Comment:
Can you add a comment on what (and why) we're updating the suffic on left and prefix on right?